### ISSUE TITLE
LZF filter: example.c seems to have a bug

### ISSUE THREAD

**Issue by Apollo3zehn** [h5py/h5py#2296]
First some environmental data:

- Operating System: Ubuntu 23.04
- Python version: -
- h5py version: 3.9.0
- HDF5 version: 1.14.0

I have compiled the LZF filter to be able to test against my own implementation and I discovered that the compression code is only executed when line https://github.com/h5py/h5py/blob/6b5af4c6495bf865fee5f036122187c21fcb17d4/lzf/example.c#L99
is executed. I guess that by closing the dataset, the chunk cache is being flushed to disk which requires the compressor to be executed.

However, `example.c` does not only write to the dataset but also reads from it right after writing. Since the chunk is still in the chunk cache, there is no need to uncompress the data and so that code is never executed.

Closing the dataset right after writing and reopening it before reading it should fix that problem.

**Comment by takluyver** [comment]
Yes, I think you're right. That file hasn't meaningfully changed in over 10 years, so it's quite possible that it behaved correctly when it was written. PRs welcome!
